### PR TITLE
Add MPI dependency

### PR DESCRIPTION
**Pull request by QuLogic**
MPI doesn't really have some standard name for the individual language dependencies, so I decided to differentiate between them using a language suffix. OpenMPI is the only implementation to have added pkg-config files (and those are separately named) but most others still use the compiler wrappers. I tried #both OpenMPI with the pkg-config files and without, but I have not tried with any other implementations.

Alternatively, we could add some sort of `language` keyword instead of modifying the name.

I expect these will fail on CI at the moment since nothing has an MPI install. Ubuntu should be an `apt install` away; I don't know about macOS; and on Windows, it might be possible to install MS-MPI.

### CODE REVIEW

**Comment by QuLogic** [comment]
Are there any thoughts on the way language-specific libraries are specified here? Would a separate keyword argument be a better option?

**Comment by QuLogic** [comment]
This should be good now. I've tested with gcc/gfortran+openmpi/mpich and intel+openmpi/intelmpi (in older and newer versions).

@jpakkane I think would require `openmpi-bin`, `libopenmpi-dev` to be installed in the container.

**Comment by jpakkane** [comment]
CI image updated.

File: mesonbuild/dependencies/misc.py [hunk 1]
```diff
@@ -277,6 +279,211 @@ def need_threads(self):
...
+            return
+        incdir = os.environ['MSMPI_INC']
+        arch = detect_cpu_family(self.env.coredata.compilers)
+        if arch == 'x86':
+            if 'MSMPI_LIB32' not in os.environ:
+                return
+            libdir = os.environ['MSMPI_LIB32']
+            post = 'x86'
+        elif arch == 'x86_64':
+            gl = 'lib64*'
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[flake8]__
* [F841] local variable 'gl' is assigned to but never used

**Comment by QuLogic** [comment]
I added some support for MS-MPI as well, but it looks like it's failing to cross-compile. When did that start happening? Should it be skipped or moved to a different test directory?

**Comment by QuLogic** [comment]
Rebased for conflicts.

**Comment by QuLogic** [comment]
Rebased for conflicts again (using a single file for changelog is really annoying).

**Comment by jpakkane** [comment]
That will get fixed in the next release as we'll switch to having an individual snippet file for each new feature.

**Comment by jpakkane** [comment]
I'm not a fan of adding new keyword arguments unless they are absolutely necessary. Would it make sense to to have that in the name instead? So you'd have `mpi-c`, `mpi-cpp` and `mpi-fortran`. This would match ompi package names `ompi-c` etc. Would that clash with something or be unreadable? Is there ever the case that there are multiple MPI implementations available and we would need to provide a way for the user to specify which one to use?

If we do add the new keyword then the following needs to be done:

- document the new keyword in the reference manual entry on dependencies
- make it a hard error to specify the `language` keyword in any dependency that does not explicitly handle it (i.e. anything except MPI)

**Comment by QuLogic** [comment]
That's how it was done originally before @nirbheek preferred this method:
```
(06:31:58 PM) QuLogic: OK, how about the naming though? dependency('mpi:c') vs something like dependency('mpi', language : 'c')?
(06:34:21 PM) jpakkane: Does it provide pkg-config files? What are those called?
(06:34:44 PM) QuLogic: only one implementation does
(06:35:38 PM) QuLogic: those're ompi-c, ompi-cxx, etc., but the o prefix is because it's named OpenMPI
(06:36:09 PM) QuLogic: other implementations don't provide pkg-config files that I know of
(06:37:11 PM) nirbheek: QuLogic, latter is better
(06:37:34 PM) nirbheek: It should be useful for pkg-config too, since .pc files don't specify a language
```

I don't see any point in supporting multiple implementations in the same build; that seems rather specialized.

**Comment by jpakkane** [comment]
Let's do the keyword then. This just needs the two fixes mentioned above.

> I don't see any point in supporting multiple implementations in the same build

I was thinking more of having `foompi` and `barmpi` on the same system and being able to tell which one to use per build directory. That's probably quite niche too.

**Comment by QuLogic** [comment]
Made the two fixes above.

I think using to support multiple implementations, you'd want to try and search for each one explicitly instead of using this generic any-implementation dependency finder.

**Pull request merged**